warehouse/accounts/views.py

```diff
@@ -1288,3 +1295,244 @@ def reauthenticate(request, _form_class=ReAuthenticateForm):
+
+    def _check_ratelimits(self):
+        if not self._ratelimiters["user.oidc"].test(self.request.user.id):
+            raise TooManyOIDCRegistrations(
+                resets_in=self._ratelimiters["user.oidc"].resets_in(
+                    self.request.user.id
+                )
+            )
+
+        if not self._ratelimiters["ip.oidc"].test(self.request.remote_addr):
+            raise TooManyOIDCRegistrations(
+                resets_in=self._ratelimiters["ip.oidc"].resets_in(
+                    self.request.remote_addr
+                )
+            )
+
+    @property
+    def pending_github_provider_form(self):
+        return PendingGitHubProviderForm(
+            self.request.POST,
+            api_token=self.request.registry.settings.get("github.token"),
+            project_factory=self.project_factory,
+        )
+
+    @property
+    def default_response(self):
+        return {
+            "oidc_enabled": self.oidc_enabled,
+            "pending_github_provider_form": self.pending_github_provider_form,
+        }
+
+    @view_config(request_method="GET")
+    def manage_publishing(self):
+        if not self.oidc_enabled:
+            raise HTTPNotFound
+
+        return self.default_response
+
+    @view_config(
+        request_method="POST", request_param=PendingGitHubProviderForm.__params__
+    )
+    def add_pending_github_oidc_provider(self):
+        if not self.oidc_enabled:
+            raise HTTPNotFound
+
+        self.metrics.increment(
+            "warehouse.oidc.add_pending_provider.attempt", tags=["provider:GitHub"]
+        )
+
+        if self.request.flags.enabled(AdminFlagValue.DISALLOW_OIDC):
+            self.request.session.flash(
+                (
+                    "OpenID Connect is temporarily disabled. "
+                    "See https://pypi.org/help#admin-intervention for details."
+                ),
+                queue="error",
+            )
+            return self.default_response
+
+        if not self.request.user.can_register_pending_oidc_providers:
+            self.request.session.flash(
+                self.request._(
+                    "This account isn't allowed to register pending OpenID Connect "
+                    "providers. See https://pypi.org/help#openid-connect for details."
+                ),
+                queue="error",
+            )
+            return self.default_response
+
+        # Separately from having permission to register pending OIDC providers,
+        # we limit users to no more than 3 pending providers at once.
+        if len(self.request.user.pending_oidc_providers) >= 3:
+            self.request.session.flash(
+                self.request._(
+                    "You can't register any more pending OpenID Connect providers "
+                    "at the moment."
+                ),
+                queue="error",
+            )
+            return self.default_response
```

Comment:
Flagging for review: I didn't put this in the form validation logic (below) because it's a property of the user's state and not the form itself, but I could change that.